### PR TITLE
feature: add support for SetSuccessThresholdSinks(...)

### PR DESCRIPTION
**Pull request by jimlambrt**
SetSuccessThresholdSinks sets the success threshold per eventType. For the overall processing of a given event to be considered a success, at least as many sinks as the threshold value must successfully process the event.

### CODE REVIEW

File: broker.go [hunk 1]
```diff
@@ -219,3 +227,24 @@ func (b *Broker) SetSuccessThreshold(t EventType, successThreshold int) error {
...
+		return fmt.Errorf("successThresholdSinks must be 0 or greater")
+	}
+
+	g, ok := b.graphs[t]
+	if !ok {
+		g = &graph{}
+		b.graphs[t] = g
+	}
+
+	g.successThreshold = successThresholdSinks
```
Comment (peteski22) [inline broker.go:248]:
```suggestion
	g.successThresholdSinks = successThresholdSinks
```

File: graph.go [hunk 2]
```diff
@@ -17,9 +17,15 @@ type graph struct {
...
-	// successThreshold specifies how many sinks must store an event for Process
-	// to not return an error.
+	// successThreshold specifies how many pipelines must successfully process
+	// an event for Process to not return an error.  This means that a filter
+	// could of course filter an event before it reaches the pipeline's sink,
+	// but it would still count as success when it comes to meeting this threshold
 	successThreshold int
+
+	// successThresholdSinks specifies how many sinks must store an event for Process
+	// to not return an error.
```
Comment (peteski22) [inline graph.go:27]:
```suggestion
	// successThresholdSinks specifies how many sinks must successfully process 
	// an event for Process to not return an error.
```

File: graph_test.go [hunk 3]
```diff
@@ -141,62 +141,69 @@ func TestSendResult(t *testing.T) {
...
 		},
 		{
-			name:      "one bad sink with filter true threshold=1",
+			name:      "one bad sink with filter threshold=1 thresholdSink=0",
 			sinkIDs:   []NodeID{badSink},
-			threshold: 1, warnings: 0, sent: 1, failure: false, filter: true,
+			threshold: 1, thresholdSinks: 0, warnings: 0, sent: 1, completeSinks: 0, failure: false, filter: true,
+		},
+		{
+			name:      "one bad sink with filter threshold: 1 thresholdSinks=1",
```
Comment (peteski22) [inline graph_test.go:204]:
```suggestion
			name:      "one bad sink with filter threshold=1 thresholdSinks=1",
```

File: broker.go
Comment (peteski22) [inline broker.go:1]: `broker_test.go` will need a test for `SetSuccessThresholdSinks`:

```
func TestSuccessThresholdSinks(t *testing.T) {
	b := NewBroker()

	err := b.SetSuccessThresholdSinks("t", 2)
	if err != nil {
		t.Fatal(err)
	}
	g, ok := b.graphs["t"]
	if !ok {
		t.Fatalf("expected graph for eventType")
	}
	if g.successThresholdSinks != 2 {
		t.Fatalf("expected successThresholdSinks %d, not %d", g.successThresholdSinks, 2)
	}

	err = b.SetSuccessThresholdSinks("t", -1)
	if err == nil || err.Error() != "successThresholdSinks must be 0 or greater" {
		t.Fatalf("expected successThresholdSinks error")
	}
}
```

**Review by peteski22 — CHANGES_REQUESTED** [review]
Thanks for this Jim, I think there might be a little catch on a bug (added comments/suggestions), but LGTM apart from that. 😄

File: broker.go (hunk 1 above)
Comment (jimlambrt) [reply broker.go:248]: ty!

File: graph.go (hunk 2 above)
Comment (jimlambrt) [reply graph.go:27]: ty!

File: graph_test.go (hunk 3 above)
Comment (jimlambrt) [reply graph_test.go:204]: ty!

**Review by peteski22 — APPROVED** [review]
👍🏼 nice one

File: broker_test.go [hunk 4]
```diff
@@ -362,3 +362,24 @@ func TestSuccessThreshold(t *testing.T) {
...
+	err := b.SetSuccessThresholdSinks("t", 2)
+	if err != nil {
+		t.Fatal(err)
+	}
+	g, ok := b.graphs["t"]
+	if !ok {
+		t.Fatalf("expected graph for eventType")
+	}
+	if g.successThresholdSinks != 2 {
+		t.Fatalf("expected successThresholdSinks %d, not %d", g.successThresholdSinks, 2)
```
Comment (peteski22) [inline broker_test.go:378]:
```suggestion
		t.Fatalf("expected successThresholdSinks %d, not %d", 2, g.successThresholdSinks)
```

Comment (peteski22) [reply broker_test.go:378]: This is my bad as I just copy pasta the test above, I think maybe the order of the args is the wrong way around? 

If so, the L.357 also needs to be adjusted:

`t.Fatalf("expected successThreshold %d, not %d", 2, g.successThreshold)`

**Review by peteski22 — CHANGES_REQUESTED** [review]
Just one thing to eye-ball and see if you agree :)

File: broker_test.go (hunk 4 above)
Comment (jimlambrt) [reply broker_test.go]: updated both tests.

**Pull request merged**